Mathlib/RingTheory/PowerSeries/Order.lean

```diff
@@ -305,17 +328,61 @@ theorem order_X_pow (n : ℕ) : order ((X : R⟦X⟧) ^ n) = n := by
 
 end OrderZeroNeOne
 
-section OrderIsDomain
+section NoZeroDivisors
 
--- TODO: generalize to `[Semiring R] [NoZeroDivisors R]`
-variable [CommRing R] [IsDomain R]
+variable [Semiring R] [NoZeroDivisors R]
 
 /-- The order of the product of two formal power series over an integral domain
  is the sum of their orders. -/
 theorem order_mul (φ ψ : R⟦X⟧) : order (φ * ψ) = order φ + order ψ := by
-  classical
+/-   classical
   simp only [order_eq_emultiplicity_X]
-  exact emultiplicity_mul X_prime
+  exact emultiplicity_mul X_prime -/
+  apply le_antisymm _ (le_order_mul _ _)
+  by_cases h : φ.order = ⊤ ∨ ψ.order = ⊤
+  · rcases h with h | h <;> simp [h]
+  · simp only [not_or, ENat.ne_top_iff_exists] at h
+    obtain ⟨m, hm⟩ := h.1
+    obtain ⟨n, hn⟩ := h.2
+    rw [← hm, ← hn, ← ENat.coe_add]
+    rw [eq_comm, order_eq_nat] at hm hn
+    apply order_le
+    rw [coeff_mul, Finset.sum_eq_single ⟨m, n⟩]
+    · exact mul_ne_zero_iff.mpr ⟨hm.1, hn.1⟩
+    · intro ij hij h
+      by_cases h' : ij.1 < m
+      · rw [hm.2 ij.1 h', zero_mul]
+      · rw [hn.2 ij.2]
+        · rw [mul_zero]
+        · simp only [mem_antidiagonal] at hij
+          apply Nat.lt_of_add_lt_add_left
+          rw [← hij]
+          refine Nat.add_lt_add_right ?_ ij.2
+          rw [lt_iff_le_and_ne]
+          refine ⟨not_lt.mp h', ?_⟩
+          intro h''
+          apply h
+          ext
+          · exact h''.symm
+          · dsimp
+            simpa only [h'', Nat.add_right_inj] using hij
+    · intro h
+      apply False.elim (h _)
+      simp [mem_antidiagonal]
+
+theorem order_pow [Nontrivial R] (φ : R⟦X⟧) (n : ℕ) :
```

Comment:
This is true for the trivial ring, isn't it?

Reply:
oh god…

Reply:
No sorry, it doesn't work if `n=0`. Maybe we want both versions.

Reply:
It does work, indeed. because 0 * infinity = 0.